IKeyEncryptionKey::KeyId type Uri => string

jaschrep-msft opened the pull request:
Storage previously supported providers using any string as a key id. Type must be relaxed from Uri back to string to be compatible with data already in storage.

jaschrep-msft committed:
IKeyEncryptionKey::KeyId type Uri => string

schaabs commented on sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:
```diff
@@ -97,7 +97,7 @@ internal CryptographyClient(JsonWebKey keyMaterial, TokenCredential credential,
         /// The <see cref="Key.Id"/> of the key used to perform cryptographic operations for the client.
         /// </summary>
-        public Uri KeyId => _keyId;
+        public string KeyId => _keyId.ToString();
```
> public string KeyId => _keyId.ToString(); [](start = 7, length = 42)

I would rather leave the public property a Uri and explicitly implement the interface:

string IKeyEncryptionKey.KeyId { get { return _keyId.ToString(); } }

heaths replied:
How does this related to https://github.com/Azure/azure-sdk-for-net/pull/7775#discussion_r329691722? Should indeed keep it a string, or was this only true for the JWK?

schaabs commented:
In order to fully facilitate using non-uri style Key Ids you also need to update IKeyEncryptionKeyResolver.Resolve and ResolveAsync to take string keyId rather than Uri keyId.

schaabs reviewed (changes requested):
In order to fully facilitate using non-uri style Key Ids you also need to update IKeyEncryptionKeyResolver.Resolve and ResolveAsync to take string keyId rather than Uri keyId.

heaths commented:
@schaabs, but then what if those aren't URIs, which was the justification of this change? Would that imply, for example, a relative URI, say, a secret ID sans the sever/base path?

jaschrep-msft commented:
@schaabs It looks as if IKeyEncryptionKeyResolver::Resolve and ResolveAsync already take strings for the keyId.

jaschrep-msft committed:
CryptographyClient explicitly implements interface

Changed CryptographyClient::KeyId back to being of type Uri. Class now
implicitly implements the change to IKeyEncryptionKey where KeyId has
been changed to type string.

schaabs commented:
@jaschrep-msft I take back my comment about having the interface being explicitly implemented. Given we are changing this on JsonWebKey to accomidate string keys as well. The changes as you have them are fine.

jaschrep-msft commented:
Rolling back.

jaschrep-msft committed:
Revert "CryptographyClient explicitly implements interface"

This reverts commit 55f180707e0fa45f598616eaf60e15fd6a02ac50.
